presto-jdbc/src/main/java/com/facebook/presto/jdbc/PrestoDriver.java

```diff
@@ -57,30 +66,196 @@
+     *       All callers atomically increment the refcount on the queryExecutor
+     *    b) Suitable cache entry:
+     *       All callers of get() obtain the read lock
+     *       The cache returns the existing QueryExecutor to all callers of
+     *         get() in some arbitrary sequence.
+     *       All callers atomically increment the refcount on the queryExecutor
+     *
+     * 2) Simultaneous close() operations happening, no get() operation.
+     *    Each thread calling close() gets the write lock in turn and:
+     *       decrements the refcount
+     *       checks to see if the refcount is zero.
+     *       if so, invalidates the cache entry for the queryExecutor
+     *          The cache removes the queryExecutor under its own lock
+     *          and releases its own lock.
+     *       The RemovalListener is called, which closes the HTTP client
+     *       releases the write lock
+     *
+     * 3) Simultaneous get() and close(), get obtains read lock first
+     *    Caller of get() obtains read lock and:
+     *       calls cache.get() which returns existing queryExecutor under the
+     *             cache's own lock
+     *       increments queryExecutor.refcount under readlock
+     *       releases read lock
+     *    Caller of close() obtains write lock and:
+     *       decrements refcount
+     *       refcount != 0
+     *       releases write lock
+     *
+     * 4) Simultaneous close() and get(), close() obtains write lock first
+     *    Caller of close() obtains write lock and:
+     *       decrements refcount
+     *       refcount == 0
+     *       invalidates cache, as in 2 above, closing HTTP client.
+     *       releases write lock
+     *    Caller of get() obtains read lock and:
+     *       calls cache.get() which returns a new queryExecutor under the
+     *             cache's own lock
+     *       increments queryExecutor.refcount under readlock
+     *       releases read lock
+     *
+     * The last case of interest is the process that is using PrestoDriver
+     * calling prestoDriver.close(). This invalidates all of the cached
+     * QueryExecutors regardless of their refcount, and in doing so closes
+     * their HTTP clients. If anybody is still trying to uses a
+     * PrestoConnection created by the driver, they're SOL.
+     */
     public PrestoDriver()
     {
-        this.queryExecutor = QueryExecutor.create(DRIVER_NAME + "/" + DRIVER_VERSION);
+        this.cacheLock = new ReentrantReadWriteLock();
+
+        Consumer<QueryExecutor> queryExecutorCloser = new Consumer<QueryExecutor>() {
+            @Override
+            public void accept(QueryExecutor queryExecutor)
+            {
+                cacheLock.writeLock().lock();
+                if (queryExecutor.dereference() == 0) {
+                    queryExecutorCache.invalidate(queryExecutor.getClientProperties());
+                }
+                cacheLock.writeLock().unlock();
```

Comment:
Put it to finally. But frankly i would prefer intrinsic `synchronized` locking mechanism here